Support remote for builtin operators in distributed autograd

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28656 Support remote for Python UDF in distributed autograd
* **#28630 Support remote for builtin operators in distributed autograd**

This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

Differential Revision: [D18138562](https://our.internmc.facebook.com/intern/diff/D18138562)

mrshenli committed:
Support remote for builtin operators in distributed autograd

This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

[ghstack-poisoned]

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -295,16 +302,70 @@ def _test_graph(self, fn):
     def test_graph_for_python_call(self):
         self._test_graph(my_py_add)
 
+    def _test_graph_remote(self, fn):
```
Will consolidate this with the one for rpc_sync

mrshenli committed:
Update on "Support remote for builtin operators in distributed autograd"


This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for builtin operators in distributed autograd"


This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for builtin operators in distributed autograd"


This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for builtin operators in distributed autograd"


This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

[ghstack-poisoned]

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -125,7 +132,7 @@ def world_size(self):
         return INIT_METHOD_TEMPLATE.format(file_name=self.file_name)
 
-    @dist_init
+    @dist_init(setup_model_parallel=True)
```
https://github.com/pytorch/pytorch/pull/27940 will land soon, we shouldn't need to specify the default arg after that.

mrshenli replied:
Yes, I will rebase after that lands.

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -99,9 +100,15 @@ def _exec_func(self, exec_mode, method, *args):
+                              args=(args))
+            return rref.to_here().wait()
+        else:
+            raise ValueError("Unrecognized ExecMode {}".format(exec_mode))
```
This is awesome! Thanks for refactoring dist_autograd_test in such a clean way to enable RRef tests.

zhaojuanmao commented:
overall lgtm

but after this diff, making me think whether we should send autograd meta, create autograd context and check whether we should attach autograd function for every rpc call in default. that will simplify our codes a lot.

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -536,20 +576,21 @@ def _verify_backwards_remote(self, tensors, context_id, local_grads, *args):
                 ret = self._exec_func(exec_mode, torch.add, t1, t2)
                 loss = ret.sum()
-                local_grads = self._verify_backwards(exec_mode, [loss], context_id, local_grads, t1, t2)
+                ret = self._verify_backwards(exec_mode, [loss], context_id, local_grads, t1, t2)
```
Shouldn't we also verify here (and in other tests) that a gradient is accumulated on the OwnerRRef?

mrshenli replied:
The result tensor of `torch.add` (owner) actually does not require grad, but let me add another test to cover that.

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -735,12 +780,12 @@ def test_backward_invalid_args(self):
                 dist_autograd.backward([t])
 
-    @dist_init
+    @dist_init(setup_model_parallel=True)
```
Can we add the following tests:

1) Never call RRef.to_here() and verify grads are accumulated on all tensors except the RRef tensor.
2) A creates an RRef on B and shares that RRef with C. C now calls to_here() and does some computation and returns the result to A. A now calls backward on the result. We should verify the OwnerRRef on B accumulates gradients appropriately.
3) Multiple nodes call RRef.to_here() on the same rref and perform some computation. The backward pass should result in multiple edges to the same RRef and appropriate gradient accumulation. We could denote one node as the parameter server for this test and retrieve the RRef to do some computation.

pritamdamania87 commented on torch/csrc/distributed/autograd/utils.cpp:
```diff
@@ -42,37 +42,45 @@ DistAutogradContext* addRecvRpcBackward(
-      autogradMetadata, autogradContext, fromWorkerId);
-  for (auto& tensor : tensors) {
-    torch::autograd::set_history(tensor, grad_fn);
+  if (!tensors.empty()) {
```
Did we catch this issue with the unit tests you added? If not can we add a unit test?

mrshenli replied:
Yes, the `to_here()` message is wrapped with autograd, which will trigger this method on the caller. However, we should not insert a recv autograd function for it because that message does not contain any tensor.

pritamdamania87 commented on torch/csrc/distributed/autograd/utils.h:
```diff
@@ -34,19 +34,23 @@ TORCH_API DistAutogradContext* addRecvRpcBackward(
     rpc::Message&& wrappedRpcMsg,
-    rpc::MessageType msgType);
+    rpc::MessageType msgType,
+    bool checkRequiresGrad = true);
```
nit: maybe call this `forceGradRecording` and default it false. This way `ScriptRRefFetchCall` can set this to true, to force recording grad even if tensors are empty.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -133,14 +136,36 @@ template <>
-    auto rfr = ScriptRRefFetchRet::fromMessage(message);
-    future->markCompleted(rfr->values().front());
+    if (message.type() == MessageType::FORWARD_AUTOGRAD_RESP) {
+      auto response = deserializeResponse(message);
```
nit: This deserializeResponse can be outside the if statement right? Then the result can be used in either of the if-else clauses?

pritamdamania87 reviewed (changes requested):
Looks great! Mostly requesting changes for a few additional unit tests.

mrshenli committed:
Update on "Support remote for builtin operators in distributed autograd"


This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

Differential Revision: [D18138562](https://our.internmc.facebook.com/intern/diff/D18138562)

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for builtin operators in distributed autograd"


This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

Differential Revision: [D18138562](https://our.internmc.facebook.com/intern/diff/D18138562)

[ghstack-poisoned]

mrshenli committed:
Update on "Support remote for builtin operators in distributed autograd"


This includes:
1. Respect autograd context in rpc.remote for builtin ops
2. Force setting autograd context in RRef.to_here() even if the
message for to_here() does not contain any tensor.

Differential Revision: [D18138562](https://our.internmc.facebook.com/intern/diff/D18138562)

[ghstack-poisoned]
